File: proposed-projects/round-trip-experiment/reverse/20260909T133702Z_8305aa82/20260910T075701Z_f285a71f/ported_implementation/src/rules-builder/symbol-ids.ts
/** We don't need to delete, just preserve relationships. */
export class SymbolIds {
  private map: Map<string, number> = new Map();
  private pos: Map<string, number> = new Map();
  private reverseMap: Map<number, string> = new Map();

  get size(): number {
    return this.map.size;
  }

  keys(): string[] {
    return [...this.map.keys()];
  }

  has(key: string): boolean {
    return this.map.has(key);
  }

  get(key: string): number {
    if (!this.map.has(key)) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return this.map.get(key) as number;
  }

  reverseGet(key: number): string {
    if (!this.reverseMap.has(key)) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return this.reverseMap.get(key) as string;
  }

  getPos(key: string): number {
    if (!this.pos.has(key)) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return this.pos.get(key) as number;
  }

  set(key: string, value: number, pos: number): void {
    this.map.set(key, value);
    this.pos.set(key, pos);
    this.reverseMap.set(value, key);
  }

  *[Symbol.iterator](): IterableIterator<[string, number]> {
    yield* [...this.map.entries()];
  }
}

File: packages/gbnf-experiment/data/20260908T234247Z_2da5a2c0/ported_implementation/gbnf/rules_builder/symbol_ids.py
"""Port of ``src/rules-builder/symbol-ids.ts``."""

from __future__ import annotations

from typing import Dict, Iterator, Tuple


class SymbolIds:
    # we don't need to delete, just preserve relationships
    def __init__(self) -> None:
        self._map: Dict[str, int] = {}
        self._pos: Dict[str, int] = {}
        self._reverse_map: Dict[int, str] = {}

    @property
    def size(self) -> int:
        return len(self._map)

    def __len__(self) -> int:
        return len(self._map)

    def keys(self) -> Iterator[str]:
        return iter(self._map.keys())

    def has(self, key: str) -> bool:
        return key in self._map

    def __contains__(self, key: str) -> bool:
        return key in self._map

    def get(self, key: str) -> int:
        if key not in self._map:
            raise KeyError(f"SymbolIds does not contain key: {key}")
        return self._map[key]

    def reverse_get(self, key: int) -> str:
        if key not in self._reverse_map:
            raise KeyError(f"SymbolIds does not contain value: {key}")
        return self._reverse_map[key]

    def get_pos(self, key: str) -> int:
        if key not in self._pos:
            raise KeyError(f"SymbolIds does not contain key: {key}")
        return self._pos[key]

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._pos[key] = pos
        self._reverse_map[value] = key

    def __iter__(self) -> Iterator[Tuple[str, int]]:
        return iter(self._map.items())
